instrumentation/apache-httpasyncclient-4.0/src/main/java/io/opentelemetry/auto/instrumentation/apachehttpasyncclient/ApacheHttpAsyncClientDecorator.java

```diff
@@ -76,8 +77,26 @@ protected Integer status(final HttpContext context) {
   }
 
   @Override
-  protected String userAgent(HttpRequest httpRequest) {
-    final Header header = httpRequest.getFirstHeader(USER_AGENT);
+  protected String requestHeader(HttpRequest request, String name) {
+    return header(request, name);
+  }
+
+  @Override
+  protected String responseHeader(HttpContext context, String name) {
+    HttpResponse response = extractResponse(context);
+    if (response != null) {
+      return header(response, name);
+    }
+    return null;
+  }
+
+  private static String header(HttpMessage message, String name) {
+    Header header = message.getFirstHeader(name);
     return header != null ? header.getValue() : null;
   }
+
+  private static HttpResponse extractResponse(HttpContext context) {
+    Object responseObject = context.getAttribute(HttpCoreContext.HTTP_RESPONSE);
+    return responseObject instanceof HttpResponse ? (HttpResponse) responseObject : null;
+  }
```

Comment:
instead of extracting the response from context twice (even though cheap), do u mind trying out changing ApacheHttpAsyncClientDecorator generic type from HttpContext to HttpResponse, and extracting response from context in TraceContinuedFutureCallback when calling `onResponse`, e.g.

```
    public void completed(final T result) {
      HttpResponse response = getResponse(context);
      DECORATE.onResponse(clientSpan, response);
      ...
    }
```

i think this may be nicer

also, just noticed, then won't need the `response != null` conditionals above since `response != null` is already checked prior to calling those methods

Reply:
Cool, looks much nicer